File: wear/src/test/java/io/github/sirbughunter/agenticwear/data/PairingAuthTest.kt
package io.github.sirbughunter.agenticwear.data

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class PairingAuthTest {
    @Test
    fun matchesBridgeAuthenticationVector() {
        val authenticator = PairingAuthenticator.fromCode("ABCD-2345")
        try {
            assertEquals("enAbNInJ8okG3POKtCwnI0kuURH9viQTZlCPpcldao4", authenticator.pairId)
            assertEquals(
                "aXjRSCkiEwYziC4YxKbGTPV33nTzES67NNBxcHqn-wo",
                authenticator.createProof("watch", "bridge-public-key", "watch-public-key"),
            )
            assertEquals(
                "v--ZPVadRpXfaAsXsoHdnWy0tzbBMHApKFCrAMT5vow",
                authenticator.createProof("bridge", "bridge-public-key", "watch-public-key"),
            )
        } finally {
            authenticator.clear()
        }
    }

    @Test
    fun rejectsRoleAndKeySubstitution() {
        val authenticator = PairingAuthenticator.fromCode("ABCD2345")
        try {
            val proof = authenticator.createProof("watch", "bridge-public-key", "watch-public-key")
            assertTrue(authenticator.verifyProof("watch", "bridge-public-key", "watch-public-key", proof))
            assertFalse(authenticator.verifyProof("bridge", "bridge-public-key", "watch-public-key", proof))
            assertFalse(authenticator.verifyProof("watch", "changed-bridge-key", "watch-public-key", proof))
            assertFalse(authenticator.verifyProof("watch", "bridge-public-key", "changed-watch-key", proof))
            assertFalse(authenticator.verifyProof("watch", "bridge-public-key", "watch-public-key", "A".repeat(43)))
        } finally {
            authenticator.clear()
        }
    }
}

File: bridge/test/pairing-auth.test.ts
import { describe, expect, it } from "vitest";
import {
  formatPairingCode,
  generatePairingCode,
  normalizePairingCode,
  PairingAuthenticator,
} from "../src/pairing-auth.js";

const bridgePublicKey = "bridge-public-key";
const watchPublicKey = "watch-public-key";

describe("PairingAuthenticator", () => {
  it("matches the shared bridge/watch authentication vector", async () => {
    const authenticator = await PairingAuthenticator.fromCode("ABCD-2345");
    expect(authenticator.pairId).toBe("enAbNInJ8okG3POKtCwnI0kuURH9viQTZlCPpcldao4");
    expect(await authenticator.createProof("watch", bridgePublicKey, watchPublicKey))
      .toBe("aXjRSCkiEwYziC4YxKbGTPV33nTzES67NNBxcHqn-wo");
    expect(await authenticator.createProof("bridge", bridgePublicKey, watchPublicKey))
      .toBe("v--ZPVadRpXfaAsXsoHdnWy0tzbBMHApKFCrAMT5vow");
  });

  it("binds proofs to the endpoint role and both public keys", async () => {
    const authenticator = await PairingAuthenticator.fromCode("ABCD2345");
    const proof = await authenticator.createProof("watch", bridgePublicKey, watchPublicKey);
    expect(await authenticator.verifyProof("watch", bridgePublicKey, watchPublicKey, proof)).toBe(true);
    expect(await authenticator.verifyProof("bridge", bridgePublicKey, watchPublicKey, proof)).toBe(false);
    expect(await authenticator.verifyProof("watch", "changed-bridge-key", watchPublicKey, proof)).toBe(false);
    expect(await authenticator.verifyProof("watch", bridgePublicKey, "changed-watch-key", proof)).toBe(false);
    expect(await authenticator.verifyProof("watch", bridgePublicKey, watchPublicKey, "A".repeat(43))).toBe(false);
  });

  it("generates and formats unambiguous eight-character codes", () => {
    for (let index = 0; index < 32; index += 1) {
      const code = generatePairingCode();
      expect(code).toMatch(/^[A-HJ-NP-Z2-9]{8}$/u);
      expect(normalizePairingCode(formatPairingCode(code))).toBe(code);
    }
  });
});
